src/core/editor/pdf_editor.js

```diff
@@ -0,0 +1,649 @@
+      obj,
+      mustClone,
+      xref,
+      this.#getOldToNewMapping(xref)
+    );
+  }
+
+  /**
+   * Create new annotation references.
+   * @param {Array<Ref>} annotations
+   * @param {XRef} xref
+   * @returns {Promise<Array<Ref>|null>}
+   */
+  async #makeAnnotations(annotations, xref) {
+    if (!annotations || annotations.length === 0) {
+      return null;
+    }
+    const oldToNewMapping = this.#getOldToNewMapping(xref);
+    if (!oldToNewMapping.has(annotations[0])) {
+      return this.#collectDependencies(annotations, true, xref);
+    }
+    const newAnnotations = [];
+    for (const annotRef of annotations) {
+      const annotation = this.xref[oldToNewMapping.get(annotRef).num];
+      const newRef = this.newRef;
+      oldToNewMapping.put(annotRef, newRef);
+      this.xref[newRef.num] = await this.#collectDependenciesHelper(
+        annotation,
+        true,
+        xref,
+        oldToNewMapping
+      );
+      newAnnotations.push(newRef);
+    }
+    return newAnnotations;
+  }
+
+  /**
+   * @typedef {Object} pageInfo
+   * @property {PDFDocument} document
+   * @property {Set<number>} [pageIndices]
+   * @property {Array<Array<number>>} [pageRanges]
+   * @property {Set<number>} [pageDeleted]
+   */
+
+  /**
+   * Extract pages from the given documents.
+   * @param {Array<pageInfo>} pageInfos
+   * @return {Promise<void>}
+   */
+  async extractPages(pageInfos) {
+    const promises = [];
+    let newIndex = 0;
+    let pageGroupIndex = 0;
+    this.hasSingleFile = pageInfos.length === 1;
+    for (const {
+      document,
+      keptIndices,
+      keptRanges,
+      deletedIndices,
+      deletedRanges,
+    } of pageInfos) {
+      if (!document) {
+        continue;
+      }
+      promises.push(this.#collectDocumentData(document));
+      const groupIndex = pageGroupIndex++;
+      for (let i = 0, ii = document.numPages; i < ii; i++) {
+        if (deletedIndices?.has(i)) {
+          continue;
+        }
+        if (deletedRanges) {
+          let isDeleted = false;
+          for (const [start, end] of deletedRanges) {
+            if (i >= start && i <= end) {
+              isDeleted = true;
+              break;
+            }
+          }
+          if (isDeleted) {
+            continue;
+          }
+        }
+
+        let takePage = false;
+        if (keptIndices) {
+          takePage = keptIndices.has(i);
+        }
+        if (!takePage && keptRanges) {
+          for (const [start, end] of keptRanges) {
+            if (i >= start && i <= end) {
+              takePage = true;
+              break;
+            }
+          }
+        }
+        if (!takePage && !keptIndices && !keptRanges) {
+          takePage = true;
+        }
+        if (!takePage) {
+          continue;
+        }
+        const newPageIndex = newIndex++;
+        promises.push(
+          document.getPage(i).then(page => {
+            this.oldPages[newPageIndex] = new NewPageData(page, groupIndex);
```

Comment:
Since this concerns a new page, shouldn't we use `this.newPages` here instead? It's not really clear to me why we store the pages in old and then copy them to new; would it not work to only have a "new" array where we store the entries and overwrite them with a modified copy in-place?